### ISSUE TITLE
FKM of the DQ_WholeBody does not return intermediate transformations of one of the chains [BUG]

### ISSUE THREAD

**Issue by StUrb** [dqrobotics/cpp#18]
**Bug description**
Hello, @bvadorno,

I am currently working with the DQ_WholeBody  serializing two legs in order to perform a walking algorithm. One of my steps is to calculate the center of mass of the serialized chain. My approach for such calculation starts with the intermediate transformations of each link of the serialized chain. However, when I use functionality fkm(q,joint) it does not work properly.

**To Reproduce**
I used the code in the two_legs_example to create a minimal example.

*Code*

```
    % Include the DQ namespace to use i_ instead of DQ.i, etc.
    include_namespace_dq
    
    % The legs are composed of two KUKA LWR4 robots coupled together    
    right_leg = KukaLwr4Robot.kinematics();
    right_leg.name = 'Right leg';
    left_leg = KukaLwr4Robot.kinematics();
    left_leg.name = 'Left leg';
    
   
    % The left foot will be located 20 cm to the left, along the x axis,
    % but with the z-axis pointing downwards.
    left_foot = (1 - E_ * 0.5 * 0.2 * i_) * i_;
    % The right foot will be located 20 cm to the right, along the x axis,
    % but with the z-axis pointing downwards.
    right_foot = (1 + E_ * 0.5 * 0.2 * i_) * i_;
    
    
    % Physically place the left leg in the workspace
    x_0_to_left_effector = left_leg.fkm(zeros(7,1));
    new_base_frame = left_foot .* T(x_0_to_left_effector);
    left_leg.set_base_frame(new_base_frame);
    % Physically place the right leg in the workspace
    x_0_to_right_effector = right_leg.fkm(zeros(7,1));
    new_base_frame = right_foot .* T(x_0_to_right_effector);
    right_leg.set_base_frame(new_base_frame);
    
    %% Initializes the robot serially coupled kinematic chain from the mobile base
    left_to_right = DQ_WholeBody(left_leg, 'reversed');
    % The base frame is the left foot. The reference frame, also located at
    % the left foot, provides the transformation with respect to the global
    % frame. Therefore, any element in the chain will be given with respect
    % to the global frame.
    left_to_right.set_reference_frame(left_foot);
    left_to_right.set_base_frame(left_foot);
    
    % Determines the rigid transformation between the base frames of the
    % two legs
    base_left_to_base_right = left_leg.base_frame'*right_leg.base_frame;
    % Add this constant transformation to the chain
    left_to_right.add(base_left_to_base_right);    
    % Finally, add the right frame
    left_to_right.add(right_leg);
    
    q = zeros(14,1);
    x1 = left_to_right.fkm(q,1)
    x2 = left_to_right.fkm(q,2)
    x3 = left_to_right.fkm(q,3)
    x4 = left_to_right.fkm(q,4)
    x5 = left_to_right.fkm(q,5)
    x6 = left_to_right.fkm(q,6)
```
*Output*
```
x1 = 
         (1i) + E*(0.1 + 0.55j)
x2 = 
         (1i) + E*( - 0.1 + 0.55j)
x3 = 
         (1i) + E*( - 0.1)
Index exceeds matrix dimensions.

Error in DQ_WholeBody/raw_fkm (line 170)
                if isa(obj.chain{i}, 'DQ_Kinematics')

Error in DQ_WholeBody/fkm (line 142)
                x = obj.reference_frame * raw_fkm(obj,q,ith);

Error in issue_FKM_minimal_example (line 52)
    x4 = left_to_right.fkm(q,4)
```

**Expected behavior**
- I was expecting the frames of each joint of the first chain (left leg in the minimal example) 

*Expected output*
This is the output of the left leg frames without serialization.
```
x1 = 
         (0.70711 + 0.70711i) + E*(0.1096j + 0.1096k)
x2 = 
         (1) + E*(0.155k)
x3 = 
         (0.70711 - 0.70711i) + E*( - 0.25102j + 0.25102k)
x4 = 
         (1) + E*(0.355k)
x5 = 
         (0.70711 + 0.70711i) + E*(0.38891j + 0.38891k)
x6 = 
         (1) + E*(0.55k)
```

**Environment:**
 - OS: [Ubuntu 16.04]
 - dqrobotics last pull date: 28/04/2019
 - MATLAB version [R2016b]

**Comment by bvadorno** [comment]
@StUrb , just to let you know, I've found and fixed a bug in the partial fkm for reverse chains. Please be sure that you have pulled the last version.

Bruno

**Comment by StUrb** [comment]
Hey @bvadorno!

Thank you! :)

Stella

**Comment by mmmarinho** [comment]
Hello, @bvadorno ,

Can we resume this discussion? 

The `fkm()` method is a bit weird because each derived class of `DQ_Kinematics` is implementing it in a different way and I think it should be unified.

For example, in the following script we have three different behaviors:
```matlab
clear all;
close all;

robot_a = KukaLwr4Robot.kinematics();
robot_b = RoombaCreateRobot.kinematics();
robot_c = LittleJohnRobot.kinematics();


robot_vector = {robot_a, robot_b, robot_c};

for i=1:length(robot_vector)
    current_robot = robot_vector{i};
    q_size = current_robot.get_dim_configuration_space();
    q = zeros(1, q_size);
    for j=1:q_size
        current_robot.fkm(q, j);
    end
end
```

1. `robot_a`, which is a `DQ_SerialManipulator` will return the `fkm()` of each link,
2. `robot_b` which is a `DQ_DifferentialDriveRobot` will return an error because the method is not implemented with that signature. This is not a particular problem but something we should consider adding.
3. **(MAIN PROBLEM)** and `robot_c` which is a `DQ_WholeBody` will also return an error because it returns a `DQ_Kinematics` object instead of a `DQ`. <-- This is the main problem for me to implement in other languages.

So for now what I suggest is that we unify the `fkm()` as something like
- `DQ fkm(q, to_ith_link)`.

I think that `DQ_WholeBody` doesn't need this special `fkm()` because it already has the method `get_chain(ith)` to return the ith `DQ_Kinematics`.

What do you think? This is kind of a big deal so I would try to add it to a revision of this release, let's say 19.0.1 ASAP.

**Comment by bvadorno** [comment]
Hi @mmmarinho,

Since FKM is a widely used concept in robotics, I don't think it's a good idea to remove it from DQ_WholeBody. Furthermore, I think that both signatures are useful, namely `fkm(q)` and `fkm(q,ith_link)`, where the former is equivalent to `fkm(q,last_link)`.

We could add the second signature to  `DQ_DifferentialDriveRobot`, although it would be useless because `fkm(q)` and `fkm(q,ith_link)` would return always the same thing because `ith_link` always equals 1. Even though I'm not bothered by the fact that it currently does not implement the second signature, I don't have any strong objection to implementing it for the sake of uniformity.

In the case of `DQ_WholeBody`, I think I didn't get your point. Actually, `fkm` does return a `DQ` object, which is the expected behavior of such a method. On the other hand, `get_chain` returns a `DQ_Kinematics` object, which is also its intended behavior. That's why we need both. In your description, however, you say that `fkm` returns a `DQ_Kinematics` object, which is odd indeed. Are you referring to the Matlab implementation? Could you please clarify this issue?

On the other hand, I agree that the signature with three parameters in `DQ_WholeBody` does not follow the same behavior as the other methods. Maybe we could eliminate the third parameter, but I'm not sure what would be the impact on the implementation and overall usage. If we really think that it would be an improvement, I'd rather implement it in our next release to prevent further delay in our other developments.

Cheers,
Bruno

**Comment by mmmarinho** [comment]
Hello, @bvadorno ,

I think too many things got mixed up in the initial comment so let’s break it down into smaller issues first.

1. **We need to unify `fkm(q, ith)` for all derived classes of `DQ_Kinematics()`**
Main point: *Given that all derived classes are inheriting `fkm()` from `DQ_Kinematics`, it doesn’t make sense for the derived classes to have different behaviors for the same method.*

As the example I sent shows, using the same method for three derived classes of `DQ_Kinematics` has three different behaviors and that is weird for the users. I don’t think anything has to be removed, just that the behavior should be the same for all the derived classes.

The contrast is easier to see for these two derived classes:
1. In `DQ_SerialManipulator`, in `fkm(q, ith)`, the `ith` stands for a joint,
2. In `DQ_WholeBody`, in `fkm(q, ith)`, the `ith` stands for a `DQ_Kinematics` in the chain.

Suppose we want to show the frames of each of the links of a given `DQ_Kinematics` using VREP or plot it in MATLAB. The following loop would be a reasonable way of looping through each of the joints of each robot if the signature of the `fkm()` method was the same for all derived classes. But the way it is now, it won’t work.

```matlab 

clear all;
close all;

robot_a = KukaLwr4Robot.kinematics();
robot_c = LittleJohnRobot.kinematics();

robot_vector = {robot_a, robot_c};

for i=1:length(robot_vector)
    current_robot = robot_vector{i};
    q_size = current_robot.get_dim_configuration_space();
    q = zeros(1, q_size);
    for j=1:q_size
        x_of_current_joint = current_robot.fkm(q, j);
        %use x to show the frame somewhere
    end
end
```

With the current implementation, we have to specialize this loop for each derived class of `DQ_Kinematics` and that somehow beats the purpose of having the inheritance. 

2. **When?**
*I think that the sooner the better, because if we don’t change it now,  the inertia will most likely make us never change this because it will impact a lot of user-developed code*.

**Comment by bvadorno** [comment]
Thanks for clarifying your point, @mmmarinho. 

Agreed. Since this change will affect not only the `fkm` but also the `pose_jacobian` method, I'll think about a solution that affects the least as possible the current implementation and I'll let you know as soon as I finish it.

Cheers,
Bruno

**Comment by bvadorno** [comment]
Hi, @mmmarinho,
 
I created a new `DQ_SerialWholeBody` class that has methods `fkm()` and `pose_jacobian()` that work in compliance with the overall behavior of the `DQ_Kinematics` class. I also changed the class `DQ_HolonomicBase` to accept indices in those aforementioned methods to also comply with the expected behavior of the `DQ_Kinematics` class and its subclasses.

Note, however, that those indices in `DQ_HolonomicBase` do not return intermediate values in the chain because there is only one element in the chain (i.e., it's just a rigid body), despite the fact that we describe it using three generalized coordinates. Therefore, `fkm(q)` and `fkm(q,i)`, with `i ≤ 3`  return the same value for objects of  the `DQ_HolonomicBase` class. The same applies to the `jacobian()` method.

The main strategy that I adopted when modifying the class `DQ_SerialWholeBody` was to rename the old methods `fkm` and `pose_jacobian` to `fkm_by_chain` and `pose_jacobian_by_chain`, which offer the option of retrieving intermediate values by using two indices, and then I wrote new `fkm` and `pose_jacobian` methods that work as wrappers for those `xyz_by_chain` methods.

This way, if a serial robot is composed of two serially attached subchains, the first one with 5 links and the second one with 7 links, `fkm(q,8)` is equivalent to `fkm_by_chain(q,2,3)`. I also implemented the method `retrieve_chain_from_link` that finds the two indices used in `fkm_by_chain` that correspond to the index used in `fkm(q,8)`.

I'll unassign myself from this issue as now it is your turn to implement those changes in C++, but please feel free to ask for clarifications, if necessary. 

Cheers,
Bruno

**Comment by mmmarinho** [comment]
I forgot to close this issue. It has been addressed in 20.04.
